source/I2C.cpp

```diff
@@ -15,120 +15,323 @@
  */
 #include "mbed-drivers/I2C.h"
 #include "minar/minar.h"
+#include "ualloc/ualloc.h"
+#include "core-util/CriticalSectionLock.h"
+#include "PeripheralPins.h"
+
+#if DEVICE_I2C && DEVICE_I2C_ASYNCH
 
-#if DEVICE_I2C
 
 namespace mbed {
+namespace detail {
+// TODO: Increase the size of I2COwners to accept Resource Mangers for non-onchip I2C masters
+I2CResourceManager * I2COwners[MODULES_SIZE_I2C] = {nullptr};
 
-I2C *I2C::_owner = NULL;
+// use a class instead of a structure so that we get the copy constructor for free.
 
-I2C::I2C(PinName sda, PinName scl) :
-#if DEVICE_I2C_ASYNCH
-                                     _irq(this), _usage(DMA_USAGE_NEVER),
+#ifdef YOTTA_CFG_MBED_DRIVERS_I2C_POOL_ELEMENTS
+#define I2C_TRANSACTION_POOL_ELEMENTS YOTTA_CFG_MBED_DRIVERS_I2C_POOL_ELEMENTS
+#else
+#define I2C_TRANSACTION_POOL_ELEMENTS (MODULES_SIZE_I2C * 4)
 #endif
-                                      _i2c(), _hz(100000) {
-    // The init function also set the frequency to 100000
-    i2c_init(&_i2c, sda, scl);
 
-    // Used to avoid unnecessary frequency updates
-    _owner = this;
-}
+#define I2C_TRANSACTION_POOL_SIZE (I2C_TRANSACTION_POOL_ELEMENTS * sizeof(I2CTransaction))
 
-void I2C::frequency(int hz) {
-    _hz = hz;
+mbed::util::PoolAllocator I2CResourceManager::TransactionPool(
+    mbed_ualloc(I2C_TRANSACTION_POOL_SIZE,
+        {.flags=UALLOC_TRAITS_NEVER_FREE}),
```

Comment:
If we're going to end up forbidding queueing transaction from interrupt contexts, we should use the regular heap as opposed to the never-free heap.